File: LeetCode/1060.Missing_Element_in_Sorted_Array/Mario/1060.Missing_Element_in_Sorted_Array.py
#!/usr/bin/env python3


class Solution:
    def missingElement(self, nums: List[int], k: int) -> int:
        n = len(nums)

        # count how many missing elements before nums[i]
        def missing(i):
            # nums[i] - nums[0] + 1 - (i+1)
            return nums[i] - nums[0] - i

        # how many missing elements in nums
        if k > missing(n - 1):
            return nums[-1] + k - missing(n - 1)

        # so result in the [nums[0], nums[-1]] range
        left, right = 0, n - 1
        while left < right:  # left = right = n-1
            mid = (left + right) // 2
            if missing(mid) < k:
                left = mid + 1
            else:
                right = mid

        # cur left make nums[left-1] < k <= nums[left]
        return nums[left - 1] + k - missing(left - 1)

        # kth smallest / largest -> (heap / binary search)

File: LeetCode/1060.Missing_Element_in_Sorted_Array/1060.Missing_Element_in_Sorted_Array.swift
class Solution {
    func missingElement(_ nums: [Int], _ k: Int) -> Int {
        // compare missing num count, binary search
        let n = nums.count

        // [4, 7, 9, 10]
        // i = 2 nums[2] = 9, 9 - 4 - 2 = 3 (5, 6, 8)
        // how many num missing at index i
        func missing(_ i: Int) -> Int {
            return nums[i] - nums[0] - i
        }

        // last >= k
        var l = 0, r = n
        while l < r {
            let m = (l + r) / 2
            if missing(m) < k {
                l = m + 1
            } else {
                r = m
            }
        }

        return nums[l - 1] + k - missing(l - 1)
    }
}
